### PR TITLE
ClusterConnectionProvider: Allow using internal apiserver address

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

    This change makes us not use the external apiserver address when
    components running in the seed connect to the user cluster, saving one
    unnecessary hop

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
none
```

### CODE REVIEW

**Comment by mrIncompetent** [comment]
The only consumer of this functionality would be the addon controller?

**Comment by alvaroaleman** [comment]
> The only consumer of this functionality would be the addon controller?

No, the whole cluster controller

File: api/pkg/cluster/client/client.go [hunk 1]
```diff
@@ -22,19 +23,30 @@ import (
...
-	s, err := p.secretLister.Secrets(c.Status.NamespaceName).Get(resources.AdminKubeconfigSecretName)
+	var s *corev1.Secret
+	var err error
+	if p.useExternalAddress {
+		// Load the admin kubeconfig secret, it uses the external apiserver address
+		s, err = p.secretLister.Secrets(c.Status.NamespaceName).Get(resources.AdminKubeconfigSecretName)
+	} else {
+		// Load the UserClusterControllerKubeconfigSecret, it has cluster-admin privileges
+		// and uses the internal apiserver address
+		s, err = p.secretLister.Secrets(c.Status.NamespaceName).Get(resources.UserClusterControllerKubeconfigSecretName)
```
Comment (mrIncompetent) [inline api/pkg/cluster/client/client.go]: This is not explicit enough.

Clearly separate this by using 2(4) dedicated functions: 
- `GetExternalAdminKubeconfig` (`GetExternalClient`)
- `GetInternalAdminKubeconfig` (`GetInternalClient`)

Also do not reuse the UCC Kubeconfig. Instead introduce a dedicated internal admin kubeconfig which then can be used by the UCC.

File: config/kubermatic/templates/kubermatic-api-dep.yaml [hunk 2]
```diff
@@ -1,3 +1,4 @@
+{{ if .Values.kubermatic.isMaster }}
```
Comment (mrIncompetent) [inline config/kubermatic/templates/kubermatic-api-dep.yaml]: Move this into a dedicated PR

File: api/pkg/cluster/client/client.go (hunk 1 above)
Comment (alvaroaleman) [reply api/pkg/cluster/client/client.go]: > Also do not reuse the UCC Kubeconfig. Instead introduce a dedicated internal admin kubeconfig which then can be used by the UCC.

SGTM

> Clearly separate this by using 2(4) dedicated functions:

I disagree here, because that would double the whole Interface as we would also have every client getter that is part of it doubled. There is no case I can think of where someone wants to have some internal and some external clients.

What about replacing the `New` with `NewInternal` and `NewExternal` instead?

Comment (mrIncompetent) [reply api/pkg/cluster/client/client.go]: > What about replacing the `New` with `NewInternal` and `NewExternal` instead?

Great idea!

**Comment by mrIncompetent** [comment]
Doesn't this break local development?

**Comment by alvaroaleman** [comment]
> Doesn't this break local development?

Oh it indeed does. What about just adding a flag to the cluster-controller?

**Comment by mrIncompetent** [comment]
> > Doesn't this break local development?

> Oh it indeed does. What about just adding a flag to the cluster-controller?

sgtm

**Comment by alvaroaleman** [comment]
@mrIncompetent Updated, PTAL

File: api/cmd/kubermatic-controller-manager/options.go [hunk 3]
```diff
@@ -93,6 +94,7 @@ func newControllerRunOptions() (controllerRunOptions, error) {
 	flag.StringVar(&c.inClusterPrometheusScrapingConfigsFile, "in-cluster-prometheus-scraping-configs-file", "", "The file containing the custom scraping configs for the prometheus running in the cluster-foo namespaces.")
 	flag.StringVar(&c.monitoringScrapeAnnotationPrefix, "monitoring-scrape-annotation-prefix", "monitoring.kubermatic.io", "The prefix for monitoring annotations in the user cluster. Default: monitoring.kubermatic.io -> monitoring.kubermatic.io/port, monitoring.kubermatic.io/path")
 	flag.StringVar(&rawFeatureGates, "feature-gates", "", "A set of key=value pairs that describe feature gates for various features.")
+	flag.BoolVar(&c.runOutsideOfSeed, "run-outside-of-seed", false, "Whether the controller runs outside of the seed cluster")
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-controller-manager/options.go]: Would it be possible to just check if `-kubeconfig` is set?
I can't imagine any scenario where someone would set `-kubeconfig=/some/kubeconfig` but `-run-outside-of-seed=false` or vice versa

Comment (alvaroaleman) [reply api/cmd/kubermatic-controller-manager/options.go]: smart!

Comment (alvaroaleman) [reply api/cmd/kubermatic-controller-manager/options.go]: Updated, PTAL

File: api/cmd/kubermatic-controller-manager/controllers.go [hunk 4]
```diff
@@ -122,14 +122,21 @@ func createClusterController(ctrlCtx *controllerContext) (runner, error) {
 		return nil, fmt.Errorf("failed to load ImagePullSecret from %s: %v", ctrlCtx.runOptions.dockerPullConfigJSONFile, err)
 	}
 
+	var clientProvider *client.Provider
```
Comment (floreks) [inline api/cmd/kubermatic-controller-manager/controllers.go]: `clientProvider` creation logic is identical in 3 places in this file. Is it time to refactor it to a method?

**Comment by floreks** [comment]
Basically, code LGTM. I'll quickly test it.

**Comment by floreks** [comment]
Tested on my local kubermatic setup and didn't see any errors related to controller trying to go out of the cluster and back in :slightly_smiling_face:. Waiting for resolving conflicts.

**Comment by alvaroaleman** [comment]
@floreks As long as you are fine with the idea and approach of the change, you can just `/approve`, that will stay even throught new pushes. Henrik will do the final `\lgtm` (which gets removed when pushing new changes)
